crates/engine/tree/src/tree/payload_validator.rs

```diff
@@ -213,16 +213,36 @@ where
         Evm: ConfigureEngineEvm<T::ExecutionData, Primitives = N>,
     {
         match input {
-            BlockOrPayload::Payload(payload) => Ok(Either::Left(
-                self.evm_config
+            BlockOrPayload::Payload(payload) => {
+                let (iter, convert) = self
+                    .evm_config
                     .tx_iterator_for_payload(payload)
                     .map_err(NewPayloadError::other)?
-                    .map(|res| res.map(Either::Left).map_err(NewPayloadError::other)),
-            )),
+                    .into();
+
+                let iter = Either::Left(iter.into_iter().map(Either::Left));
+                let convert = move |tx| {
+                    let Either::Left(tx) = tx else { unreachable!() };
+                    convert(tx).map(Either::Left).map_err(Either::Left)
+                };
+
+                Ok((
+                    iter,
+                    Box::new(convert)
+                        as Box<
+                            dyn Fn(Either<_, _>) -> Result<Either<_, _>, _> + Send + Sync + 'static,
+                        >,
+                ))
+            }
             BlockOrPayload::Block(block) => {
```

Comment:
since this step looks slightly confusing, I think some additional docs would be helpful here